kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -355,236 +354,197 @@ uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId() const {
-  uint64_t private_start = getValueFromLatestTable(client_id_ + "_oldest");
-  uint64_t public_end = getValueFromLatestTable(kPublicEgIdKeyNewest);
-  uint64_t private_end = getValueFromLatestTable(client_id_ + "_newest");
-  LOG_DEBUG(logger_,
-            "Reading event group range, public_start: " << public_start << " private_start: " << private_start
-                                                        << " public_end: " << public_end
-                                                        << " private_end: " << private_end);
-  if (!public_start && !private_start) {
+void KvbAppFilter::readEventGroups(EventGroupId external_eg_id_start,
+                                   const std::function<bool(KvbFilteredEventGroupUpdate &&)> &process_update) {
+  uint64_t newest_public_eg_id = getNewestPublicEventGroupId();
+  uint64_t newest_private_eg_id = getValueFromLatestTable(client_id_ + "_newest");
+  uint64_t oldest_external_eg_id = oldestTagSpecificPublicEventGroupId();
+  uint64_t newest_external_eg_id = newestTagSpecificPublicEventGroupId();
+  if (not oldest_external_eg_id) {
     std::stringstream msg;
     msg << "Event groups do not exist for client: " << client_id_ << " yet.";
     LOG_ERROR(logger_, msg.str());
     throw std::runtime_error(msg.str());
   }
-  // update the offsets if we now have corresponding public/private event groups in storage
-  if (eg_data_state_->public_offset == 0) eg_data_state_->public_offset = public_start;
-  if (eg_data_state_->private_offset == 0) eg_data_state_->private_offset = private_start;
 
-  uint64_t event_group_id_end = private_end + public_end;
+  if (external_eg_id_start == 0) {
+    throw InvalidEventGroupId(external_eg_id_start);
+  }
+  if (external_eg_id_start < oldest_external_eg_id || external_eg_id_start > newest_external_eg_id) {
+    // TODO: Change exception to (id, begin, end)
+    throw InvalidEventGroupRange(external_eg_id_start, oldest_external_eg_id);
+  }
+
+  auto [global_eg_id, is_previous_private, private_eg_id, public_eg_id] = findGlobalEventGroupId(external_eg_id_start);
+  uint64_t ext_eg_id = external_eg_id_start;
+
+  uint64_t next_pvt_eg_id = private_eg_id + 1;
+  auto next_pvt_key = client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(next_pvt_eg_id);
+  uint64_t next_pub_eg_id = public_eg_id + 1;
+  auto next_pub_key = kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(next_pub_eg_id);
 
-  if (event_group_id_start == 0) {
-    throw InvalidEventGroupId(event_group_id_start);
+  // The next public or private event group might not exist or got pruned
+  // In this case, set to max so that the comparion will be lost later
+  uint64_t pvt_external_id;
+  uint64_t pvt_global_id;
+  try {
+    std::tie(pvt_global_id, pvt_external_id) = getValueFromTagTable(next_pvt_key);
+  } catch (const std::exception &e) {
+    pvt_global_id = std::numeric_limits<uint64_t>::max();
   }
```

Comment:
I would suggest changing the interface of `getValueFromTagTable()` such that we can distinguish between storage errors and lack of a key-value. That way we could maybe assert in places where it makes sense and not swallow  potentially critical errors. Maybe we can consider making `getValueFromTagTable()` return an optional.

On a side note, we can make `getValueFromTagTable()` assert if the value is not of type `ImmutableValue`. That will further simplify the interface and allow us to test less.

Reply:
We could do above (if at all) in a separate PR/commit, definitely not critical now.

Reply:
Thanks for this comment and I agree. Unfortunately, I cannot express "catch all but storage issues". I can list the current possible exceptions but as soon as this changes we will forget to update this try-catch. Asserting on the type makes sense to me - meaning, if the assert fires something is seriously wrong. I will add the assert.